### DISCUSSION TITLE
Custom auth hook - Add property to model based on expanded relations

### DISCUSSION THREAD

**Discussion by bmingles** [pocketbase/pocketbase#5527] [Q&A] [answered]
My user model currently has a `roles` relation that can contain role records for multiple organizations. 

e.g. Role records look something like:
```javascript
{ org: 'xxx', name: 'admin' }
{ org: 'xxx', name: 'user' }
{ org: 'yyy', name: 'admin' }
{ org: 'yyy', name: 'user' }
```

I then have records that are owned by organizations:
```javascript
{ org: 'xxx', someProp: 'value' }
```

I'm trying to configure api rules such that a record can be updated only if the user has an admin role for the `org` owning the record.

The closest I've been able to get is something like:
`@request.auth.roles.name ?= 'admin' && @request.auth.roles.org.id ?= org`

but I think this will actually match a record for org `yyy` if a user has an admin role in any org and they have any role in org `yyy`:

```javascript
{ org: 'xxx', name: 'admin' }
{ org: 'yyy', name: 'user' }
```

Is there a way to say match if the user has a role with `name == 'admin' && org.id == org`?

If not, are there any recommendations?

I attempted to use the `onRecordAuthRequest` hook to set a custom property `isAdmin` based on the current org + user's roles, but I couldn't figure out how to expand the relation props.

Any suggestions would be appreciated.

**Comment by ganigeorgiev** [comment, marked as answer]
If I'm understanding your description then the rule in your example should work.

Rules are evaluated together in a single run and both constraints will target the same role. You can start the executable with `--dev` and explore the generated query.

It depends for what operation the above rule is applied but in general the query should look something like:

```sql
SELECT something.*
FROM something
LEFT JOIN users on users.id = {:authID}
LEFT JOIN roles on roles.id in users.roles -- simplified fake expression
WHERE roles.name = "admin" AND roles.org = something.org -- target the same roles row
```

> I attempted to use the onRecordAuthRequest hook to set a custom property isAdmin based on the current org + user's roles, but I couldn't figure out how to expand the relation props.

I don't understand this, nor how the hook is related to the access of your other collection.

**Reply by ganigeorgiev** [reply]
_If the above doesn't help, please export your collections configuration as json (_Admin UI > Settings > Export collections_) and specify for what collection and which rule you want to configure._

**Reply by bmingles** [reply]
@ganigeorgiev thank you. The api rule worked as you suggested. I had a different issue with a required field that was causing a 400 that was conflating the permissions.

Regarding the hook question, I was trying to do something like:

```javascript
onRecordAuthRequest((event) => {
  // get org id from url or params of some sort
  const orgId = event.httpContext

  // I couldn't figure out how to get expanded roles
  const roles = event.record.expand().roles
  
  const isAdminForOrg = roles.find(
    (r) => r.org === orgId
  )

  event.record.isAdminForOrg = isAdminForOrg
})
```
